### ISSUE TITLE
Important missing bank.

### ISSUE THREAD

**Issue by zaphod77** [Wohlstand/libADLMIDI#277]
There's a FM MT-32 bank from voyetra way back in the windows 3.0 days.

There were midis composed for this bank.

The most notable of them are ballad2.mid, jazz3,mid, and reggae2.mid.  They all borked when you installed the new basic/extended/general midi windows drivers.

There were others around, but these are the only ones i still know how to find.

Sadly, i have no idea how to extract this FM bank from the old sound blaster drivers.

i'll link you a page with ballad2.mid, and a recoding of it.

https://www.vgmpf.com/Wiki/index.php?title=Ballad_2

**Comment by Wohlstand** [comment]
There are many ways, and one of them is "black box" method: I have to install such driver into DosBox, and run the MIDI file that tries every instrument, and capture the OPL output into DRO, then OPL3 Bank Editor can import these instruments. I did that method to create the SuperSapi.wopn bank file.

**Comment by zaphod77** [comment]
there is indeed a dos sb 2.0 driver that will play midis with that bank, as i recall.  but there's gotta be some easier way, i'd think.

**Comment by zaphod77** [comment]
okay i found the banks.

http://annex.retroarchive.org/cdrom/nightowl-006/042A/MID2CMF.ZIP

melody.ibk and drums.ibk in there are the actual banks.

your opl3 bank editor can take it from here.
[mt32_fm.zip](https://github.com/user-attachments/files/16407829/mt32_fm.zip)

**Comment by zaphod77** [comment]
okay, here's the bank file which will play back mt-32 midis that use defaults "correctly" on libadlmidi. created by importing mt32.ibk into the fatman 2 op patch set.

[mt32_fm.zip](https://github.com/user-attachments/files/16420324/mt32_fm.zip)

works pretty good with ballad2.mid, reggae2.mid, and jazz3.mid. :)

**Comment by Wohlstand** [comment]
Just a small question: are you sure about "volume models", and each bank has them. These "volume models" implementing different algorithms of tone to frequency conversions, and also in some cases (like the HMI model) may affect the channel allocation method (which means, will next notes played together with released decaying nodes, or they will "shut up" the previous immediately). Every WOPL Bank has the "volume model" field that means, this volume model will be always used with this bank when "auto" is selected.

**Comment by zaphod77** [comment]
seems to work fine in my testing. i'm aware it's not "perfect" but it seems close enough.  but hey if you want to actually get those old windows 3.1 drivers that play midi running, and somehow get the mt32 melody and percussion banks out of them, be my guest.  I just did what I actually knew how to do.  These files all played back properly back in the day on a single opl2 card, so i don't think they are pushing polyphony. :)

**Comment by Wohlstand** [comment]
btw, I did some another tweak, I manually repaired drums, but they are just incomplete, but I bet, Ballad now works close to original:
[MELODY.wopl.zip](https://github.com/user-attachments/files/16420570/MELODY.wopl.zip)
I guess, it's need to tweak more to figure which drums should be set.

**Comment by zaphod77** [comment]
nope. drums completely wrong.

**Comment by Wohlstand** [comment]
I do use the super-latest plain libADLMIDI. Actually, during a time I fixed rhythm mode drums as they worked totally wrong. The ADLplug seems has no updates because @jpcima didn't gave me write access to that repo until he disappeared for two years...

**Comment by zaphod77** [comment]
so it  sounds like it did on that web page when you play it?   because it didn't work, while the one i made does.

**Comment by zaphod77** [comment]
do you have a player around that will use that wopl properly?  also, do you have the test midi file that plays all instruments individually? if so i'll make the .dro file, now that i figured out how to run that old jukebox.exe

**Comment by Wohlstand** [comment]
> do you have a player around that will use that wopl properly? also, do you have the test midi file that plays all instruments individually? if so i'll make the .dro file, now that i figured out how to run that old jukebox.exe

There are right now three variants (if you use Windows according to your comments):
- The console `adlmidiplay` utility that you can build from the source from the libADLMIDI's repository (compatible with MinGW-w64 as well as with MSVC thing).
- The [VLC plugin](https://builds.wohlsoft.ru/win32/libADLMIDI-vlc-3.0.4-codec-x64.7z), it should work with latest VLC too. To open setup where you can select WOPL bank file and other options, open Options, then choose "ALL" at left-bottom corner, then through the menu find "Audio Codecs", and find "ADLMIDI" section where all the settings will be.
- The [WinMM driver](https://builds.wohlsoft.ru/win32/libADLMIDI-winmm-driver-all.7z) thing that allows you to use the control panel thing to change settings, and use regular MIDI players to play via it. However, on Windows 10 since specific revision, it gets autoremoved as "incorrectly installed" driver which makes a pain.
- The auto-build of the [Music Player](https://builds.wohlsoft.ru/win32/moondust-win64/pge-musplay-master-win64.zip) tool from the [Moondust Project](https://wohlsoft.ru/projects/Moondust/) devkit which is a gui player where you can open the MIDI setup and select the synthesizer, choose various options, bank file, and play any file you want. Right now it uses the latest version of libADLMIDI, and I update it sometimes during updates I done on the libALMIDI side including minor bugfixes or even general improvements.

**Comment by zaphod77** [comment]
okay, it does sound close, but not the same at jukebox.exe. it seems both jukebox.exe and playcmf.exe do tweak things a bit, so still need to black box the 16-bit jukebox.exe to get this right.

**Comment by Wohlstand** [comment]
About regfae2, ye, my bank is incomplete, I just attempted to put instruments by guessing without knowing how they should be placed right. It's also possible to try disassemble the thing to see the mapping thing.

**Comment by zaphod77** [comment]
yeah, easier to just send me the test midi you use, and for me to play it with jukebox.exe, and send you the .dro file.  the gm test midi should get all of them.

it's clear the mid2cmf.exe is buggy anyway, so i think that's a dead end.

**Comment by zaphod77** [comment]
still have yet to locate that test midi file so i can capture it's dro with the driver.
